### PR TITLE
feat: JSON and CSV file downloading

### PR DESCRIPTION
**Pull request by wtfua92**
- FileDownloader context with methods to download CSV and JSON files;
- Added FileDownloader menu to Header;
image

### CODE REVIEW

File: src/others/contexts/file-downloader.tsx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  const { data: aidRequests, isSuccess: aidRequestsSuccess } = useAidRequestQuery();
+  const { getActiveFilterItems } = useFilter();
+
+  const [isButtonDisabled, setButtonDisabled] = useState<boolean>(true);
+
+  const activeCategoryFilters = getActiveFilterItems("Categories");
+  const activeDateFilters = getActiveFilterItems("Dates");
+
+  useEffect(() => {
+    setButtonDisabled(!citiesSuccess && !suppliesSuccess && !aidRequestsSuccess);
```
Comment (AntonRaichuk) [inline src/others/contexts/file-downloader.tsx:58]: shouldn't it be "or" instead of &&?

File: src/others/contexts/file-downloader.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+      uniqueCategories.add(category);
+
+      const formattedRequest = {
+        category: category,
+        amount: request.requested_amount,
+      };
+
+      if (result[city]) {
+        result[city].requests.push(formattedRequest);
```
Comment (AntonRaichuk) [inline src/others/contexts/file-downloader.tsx:82]: i think there's a problem in this code with respect to "date"s.

in the filter i see that you assume that a multiple dates could be selected,
but here you are merging all requests into a dict value that has one specific date.

Comment (AntonRaichuk) [reply src/others/contexts/file-downloader.tsx:82]: let's have a nested dict? when inner one is keyed by date.
or the outer is by date and the inner by city?

File: src/index.tsx [hunk 3]
```diff
@@ -28,18 +29,20 @@ ReactDOM.render(
...
-                  <Route path="/login" element={<Login />} />
-                  <Route path="/reset-password" element={<ResetPassword />} />
-                  <Route element={<AuthWrapper />}>
-                    <Route path="/" element={<Home />} />
-                    <Route path="/requests" element={<Requests />} />
-                  </Route>
-                  <Route path="*" element={<NotFound />} />
-                </Routes>
-              </BrowserRouter>
+              <FileDownloaderContextProvider>
```
[the remark this answers is not in the document]
Comment (MaciejReimann) [reply src/index.tsx:32]: I was thinking of creating an `AidRequestcontext` - one that could hold all the fetched `AidRequest`'s in its state and expose methods like: 

- `getAsJSON()`
- `getAsCsv()`
- `filteredBy(filter: keyof AidRequest)`

That would encapsulate all the common data-processing logic related to AidRequests and made them available throughout the app in a consistent manner. WDYT?

File: src/others/contexts/file-downloader.tsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
+import { createContext, useState, useContext, FunctionComponent, useEffect } from "react";
+import { decodeCategory, decodeLocation } from "../helpers/decode-aid-request";
+import { useAidRequestQuery, useLocationsQuery, useSuppliesQuery } from "./api";
+import { useFilter } from "./filter";
+
+type FormattedAidRequests = {
+  [city: string]: {
```
Comment (MaciejReimann) [inline src/others/contexts/file-downloader.tsx:7]: **issue**

I don't think it's much needed to create yet another data shape. 

Note that here: https://github.com/ukraine-taskforce/requests-dashboard/pull/52 you have translation functions available - decodeCategory and decodeLocation is sub-optimal to say the least and should be deprecated. :)

File: src/others/contexts/filter.tsx [hunk 5]
```diff
@@ -92,10 +92,18 @@ export const FilterContextProvider: React.FunctionComponent = ({ children }) =>
     [setFilters]
   );
 
-  const getActiveFilterItems = (filterName: string) => {
+  const getActiveFilterItems = (filterName: string, field?: keyof FilterItem) => {
```
Comment (MaciejReimann) [inline src/others/contexts/filter.tsx:95]: **question**

Why csv/json requires changes to filters? I think the filters need a separate refactor (the goal would be to use `keyof AidRequest` as the only accepted filter value. 

Downloading would then require mapping filtered aid request to given format and handling the mechanics.

File: src/others/helpers/decode-aid-request.ts [hunk 6]
```diff
@@ -69,12 +69,12 @@ export const decodeAidRequestGroupedByCategoryWithTotal = (dictionary: Dictionar
   };
 };
 
-const decodeLocation = (locations: Location[], cityId: ID): DecodedLocation => {
+export const decodeLocation = (locations: Location[], cityId: ID): DecodedLocation => {
```
Comment (MaciejReimann) [inline src/others/helpers/decode-aid-request.ts:72]: As mentioned above, these should soon become deprecated.

**Review by MaciejReimann — COMMENTED** [review]
I wouldn't treat my commment as a blocker - if possible I'd deliver the feature ASAP and refactor later. That's the best strategy for the project, I believe.

**Review by MaciejReimann — APPROVED** [review]

**Comment by MaciejReimann** [comment]
Again: I believe the best strategy in the project is to deliver functionalities in the best shape possible, and refactor later. 
Let's treat my comments as non-blockers, but rather suggestions for reiterations and gradual refactor.

**Comment by wtfua92** [comment]
It was a quick POC, but if everyone is okay with what @MaciejReimann said, then let's merge it as it is and refactor it as a separate task, based already on the refactoring strategy of @MaciejReimann. 

Any objections? @Blightwidow @AntonRaichuk @jciepka? I assume that this functionality will be a PITA to refactor if we merge it post Maciej changes?

**Comment by jciepka** [comment]
> It was a quick POC, but if everyone is okay with what @MaciejReimann said, then let's merge it as it is and refactor it as a separate task, based already on the refactoring strategy of @MaciejReimann.
> 
> Any objections? @Blightwidow @AntonRaichuk @jciepka? I assume that this functionality will be a PITA to refactor if we merge it post Maciej changes?

SGTM ;)

**Pull request merged**